src/pudl/helpers.py

```diff
@@ -332,90 +332,224 @@ def is_doi(doi):
     return bool(re.match(doi_regex, doi))
 
 
-def clean_merge_asof(
+def convert_col_to_datetime(df, date_col_name):
+    """Convert a column in a dataframe to a datetime.
+
+    If the column isn't a datetime, it needs to be converted to a string type
+    first so that integer years are formatted correctly.
+
+    Args:
+        df (pandas.DataFrame): Dataframe with column to convert.
+        date_col_name (string): name of the column to convert.
+
+    Returns:
+        Dataframe with the converted datetime column.
+    """
+    if pd.api.types.is_datetime64_ns_dtype(df[date_col_name]) is False:
+        logger.warning(
+            f"{date_col_name} is {df[date_col_name].dtype} column. Converting to datetime."
+        )
+        df[date_col_name] = pd.to_datetime(df[date_col_name].astype("string"))
+    return df
+
+
+def full_timeseries_date_merge(
+    left: pd.DataFrame,
+    right: pd.DataFrame,
+    left_date_col: str = "report_date",
+    right_date_col: str = "report_date",
+    new_date_col: str = "report_date",
+    on: List[str] = [],
+    date_on: Literal["year", "quarter", "month", "day"] = "year",
+    how: Literal["inner", "outer", "left", "right", "cross"] = "inner",
+    report_at_start: bool = True,
+    freq: str = "MS",
+    **kwargs,
+):
+    """Merge dataframes with different date frequencies and expand to a full timeseries.
+
+    Arguments: see arguments for ``date_merge`` and ``expand_timeseries``
+    """
+    out = date_merge(
+        left=left,
+        right=right,
+        left_date_col=left_date_col,
+        right_date_col=right_date_col,
+        new_date_col=new_date_col,
+        on=on,
+        date_on=date_on,
+        how=how,
+        report_at_start=report_at_start,
+        **kwargs,
+    )
+    out = expand_timeseries(
+        df=out,
+        date_col=new_date_col,
+        freq=freq,
+        key_cols=on,
+    )
+    return out
+
+
+DateOnCol = Literal["year", "month", "quarter", "day"]
```

Comment:
Module level constants should `USE_ALL_CAPS` and are typically declared at the beginning of the module... oh but this is a type alias? Did you mean to use this as the type for the `date_on` arguments? They still seem to be Literals. Is that intentional?  You can also annotate constant definitions by including triple-quoted strings on the line following them.

Reply:
I wanted the type of the `date_on` column to be a list of strings where the strings must be year, month, quarter, or day. But maybe it's best if the type is just `List[str]` and there's an assert to make sure the strings in `date_on` are valid, as well as being specified in the docstring. I changed this to an assertion, let me know if you think this makes more sense.